src/openms/source/ANALYSIS/DECHARGING/MetaboliteFeatureDeconvolution.cpp

```diff
@@ -0,0 +1,1209 @@
+          double f_end1 = std::min(f1.getConvexHull().getBoundingBox().maxX(), f2.getConvexHull().getBoundingBox().maxX());
+          double f_end2 = std::max(f1.getConvexHull().getBoundingBox().maxX(), f2.getConvexHull().getBoundingBox().maxX());
+
+          double union_length = f_end2 - f_start1;
+          double intersect_length = std::max(0., f_end1 - f_start2);
+
+          if (intersect_length / union_length < rt_min_overlap)
+            continue;
+        }
+
+        // start guessing charges ...
+        mz2 = fm_out[i_RT_window].getMZ();
+
+        for (Int q1 = q_min; q1 <= q_max; ++q1) // ** q1
+        {
+          if (!chargeTestworthy_(f1.getCharge(), q1, true))
+            continue;
+
+          //DEBUG:
+          /**if (fm_out[i_RT_window].getRT()>1930.08 && fm_out[i_RT_window].getRT()<1931.2 && mz1>1443 && mz2>1443 && mz1<2848 && mz2<2848)
+          {
+            std::cout << "we are at debug location\n" << fm_out[i_RT_window].getRT() <<"   : " << mz1 << "; " << mz2 << "\n";
+          }
+          if (i_RT == 930 && i_RT_window == 931)
+          {
+            std::cout << "we are at debug location\n" << fm_out[i_RT_window].getRT() <<"   : " << mz1 << "; " << mz2 << "\n";
+          }*/
+          // \DEBUG
+
+          m1 = mz1 * q1;
+          // additionally: forbid q1 and q2 with distance greater than q_span
+          for (Int q2 = std::max(q_min, q1 - q_span + 1)
+               ; (q2 <= q_max) && (q2 <= q1 + q_span - 1)
+               ; ++q2)
+          { // ** q2
+            if (!chargeTestworthy_(f2.getCharge(), q2, f1.getCharge() == q1))
+              continue;
+
+            ++possibleEdges; // internal count, not vital
+
+            // find possible adduct combinations
+            CoordinateType naive_mass_diff = mz2 * q2 - m1;
+            double abs_mass_diff = mz_diff_max * q1 + mz_diff_max * q2; // tolerance must increase when looking at M instead of m/z, as error margins increase as well
+            hits = me.query(q2 - q1, naive_mass_diff, abs_mass_diff, thresh_logp, md_s, md_e);
+            OPENMS_PRECONDITION(hits >= 0, "MetaboliteFeatureDeconvolution querying #hits got negative result!");
+
+            // DEBUG: write out all mass values that need explanation:
+            /*if (fabs(naive_mass_diff) < 150.0)
+            {
+                if (q1 == f1.getCharge() &&
+                        q2 == f2.getCharge())
+                {
+                    dl_massdiff.push_back(naive_mass_diff - Constants::PROTON_MASS_U*	(q2-q1));
+                    il_chargediff.push_back(q2-q1);
+                }
+            }
+            if (i_RT==429 && i_RT_window==432)
+            {
+                std::cout << "DEBUG reached\n hits: " << hits << " with delta_m: " << naive_mass_diff << " and thres: " << thresh_logp << "\n";
+            }
+*/
+
+            overallHits += hits;
+            // choose most probable hit (TODO think of something clever here)
```

Comment:
one idea could be to have weighted edges, the weight could be delta-RT or a pearson or cross-correlation score of the XIC